### PR TITLE
[SPARK-13020][SQL][test] fix random generator for map type

### PR DESCRIPTION
**Pull request by cloud-fan**
when we generate map, we first randomly pick a length, then create a seq of key value pair with the expected length, and finally call `toMap`. However, `toMap` will remove all duplicated keys, which makes the actual map size much less than we expected.

This PR fixes this problem by put keys in a set first, to guarantee we have enough keys to build a map with expected length.

### CODE REVIEW

**Comment by cloud-fan** [comment]
I found this problem in a benchmark, which really confused me that map case runs faster than array case, while I thought array case and map case should have similar amount of elements.

cc @JoshRosen @nongli

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGeneratorSuite.scala [hunk 1]
```diff
@@ -93,4 +93,12 @@ class RandomDataGeneratorSuite extends SparkFunSuite {
     }
   }
 
+  test("check size of generated map") {
```
Comment (nongli) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGeneratorSuite.scala]: Is this test going to be flaky? run this in a loop 10K times as a one off

**Comment by nongli** [comment]
LGTM

File: sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGeneratorSuite.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGeneratorSuite.scala]: This test uses a fixed seed, so the test result is predictable.

**Comment by yhuai** [comment]
test this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGeneratorSuite.scala [hunk 2]
```diff
@@ -93,4 +93,12 @@ class RandomDataGeneratorSuite extends SparkFunSuite {
...
+    val mapType = MapType(IntegerType, IntegerType)
+    val generator = RandomDataGenerator.forType(mapType, nullable = false, seed = Some(13)).get
```
Comment (yhuai) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGeneratorSuite.scala]:
```
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGeneratorSuite.scala:100: not found: value seed
[error]     val generator = RandomDataGenerator.forType(mapType, nullable = false, seed = Some(13)).get
[error]                                                                            ^
```

Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGeneratorSuite.scala]: weird, I can pass it locally...

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
ping @nongli @rxin

**Comment by cloud-fan** [comment]
ping @yhuai

**Comment by cloud-fan** [comment]
retest this please

**Comment by yhuai** [comment]
LGTM. Merging to master.

**Pull request closed without merging**